plugins/chat/app/jobs/scheduled/chat/auto_join_users.rb

```diff
@@ -5,14 +5,63 @@ module Chat
     class AutoJoinUsers < ::Jobs::Scheduled
       every 1.hour
 
+      MAX_USERS = 10_000
+      LAST_SEEN_DAYS = 30
+
       def execute(_args)
         return if !SiteSetting.chat_enabled
 
-        ::Chat::Channel
-          .where(auto_join_users: true)
-          .each do |channel|
-            ::Chat::ChannelMembershipManager.new(channel).enforce_automatic_channel_memberships
-          end
+        allowed_group_permissions = [
+          CategoryGroup.permission_types[:create_post],
+          CategoryGroup.permission_types[:full],
+        ]
+
+        join_mode = ::Chat::UserChatChannelMembership.join_modes[:automatic]
+
+        sql = <<~SQL
+          WITH users AS (
+            SELECT id FROM users u
+            JOIN user_options uo ON uo.user_id = u.id
+            WHERE id > 0  AND (u.suspended_till IS NULL OR u.suspended_till <= :now)
+                          AND (u.last_seen_at IS NULL OR u.last_seen_at > :last_seen_at)
+                          AND u.active
+                          AND NOT u.staged
+                          AND uo.chat_enabled
+                          AND NOT EXISTS (SELECT 1 FROM anonymous_users a WHERE a.user_id = u.id)
+            ORDER BY last_seen_at desc
+            LIMIT :max_users
+          )
+
+          INSERT INTO user_chat_channel_memberships (user_id, chat_channel_id, following, created_at, updated_at, join_mode)
+          SELECT DISTINCT users.id AS user_id,
+                          chat_channels.id AS chat_channel_id,
+                          TRUE AS following,
+                          :now::timestamp AS created_at,
+                          :now::timestamp AS updated_at,
+                          :join_mode AS join_mode
+          FROM users
+          JOIN chat_channels on auto_join_users AND chatable_type = 'Category'
```

Comment:
Not blocking comment: We didn't have a limit previously but I wonder if we should cap the number of chat channels as well. Otherwise, the cost of this query increases with the number of auto join chat channels.

Reply:
I follow, but I prefer to hold off on this optimisation for now.